scipy/signal/_upfirdn_apply.pyx

```diff
@@ -0,0 +1,150 @@
+# LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+# NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+cimport cython
+cimport numpy as np
+import numpy as np
+
+
+def _pad_h(h, up):
+    """Store coefficients in a transposed, flipped arrangement.
+
+    For example, suppose upRate is 3, and the
+    input number of coefficients is 10, represented as h[0], ..., h[9].
+
+    Then the internal buffer will look like this::
+
+       h[9], h[6], h[3], h[0],   // flipped phase 0 coefs
+       0,    h[7], h[4], h[1],   // flipped phase 1 coefs (zero-padded)
+       0,    h[8], h[5], h[2],   // flipped phase 2 coefs (zero-padded)
+
+    """
+    h_padlen = len(h) + (-len(h) % up)
+    h_full = np.zeros(h_padlen)
+    h_full[:len(h)] = h
+    h_full = h_full.reshape(-1, up).T[:, ::-1].ravel()
+    return h_full
+
+
+class UpFIRDown(object):
+    def __init__(self, h, up, down):
+        """Helper for resampling"""
+        h = np.asarray(h, np.float64)
+        if h.ndim != 1 or h.size == 0:
+            raise ValueError('h must be 1D with non-zero length')
+        self._up = int(up)
+        self._down = int(down)
+        if self._up < 1 or self._down < 1:
+            raise RuntimeError('Both up and down must be >= 1')
+        # This both transposes, and "flips" each phase for filtering
+        self._h_trans_flip = _pad_h(h, self._up)
+
+    def _output_len(self, len_x):
+        """Helper to get the output length given an input length"""
+        return _output_len(len_x + len(self._h_trans_flip) // self._up - 1,
+                           self._up, self._down, 0)
+
+    def apply(self, x):
+        """Apply the prepared filter to a 1D signal x"""
+        out = np.zeros(self._output_len(len(x)), dtype=np.float64)
+        _apply(np.asarray(x, np.float64), self._h_trans_flip, out,
+               self._up, self._down)
+        return out
+
+
+@cython.cdivision(True)
+def _output_len(Py_ssize_t in_len,
+                Py_ssize_t up,
+                Py_ssize_t down,
+                Py_ssize_t len_h):
+    """The output length that results from a given input"""
+    cdef Py_ssize_t np
+    if len_h != 0:  # figure out what the padding needs to be
+        # the modulo below is the C-equivalent of a Python modulo,
+        # i.e. -1 % 10 = 9
+        in_len = in_len + (len_h + ((-len_h % up) + up) % up) // up - 1
+    np = in_len * up
+    cdef Py_ssize_t need = np // down
+    if np % down > 0:
+        need += 1
+    return need
+
+
+ctypedef fused DTYPE_t:
+    np.float64_t
+
+
+@cython.cdivision(True)  # faster modulo
+@cython.boundscheck(False)  # designed to stay within bounds
+@cython.wraparound(False)  # we don't use negative indexing
+cdef void _apply(DTYPE_t [:] x, DTYPE_t [:] h_trans_flip, DTYPE_t [:] out,
+                 Py_ssize_t up, Py_ssize_t down) nogil:
```

Comment:
W.r.t. axis argument: unless I'm missing something, I don't think it'll work like you say it would: I doubt `apply_along_axis` turns an n-dim array into a 1D one, and cython code expect the latter. What you could do is to roll the relevant axis to be the first one, ravel the array, feed it to cython, and then restore the shape of the result. This is what `PPoly` and friends do.